File: Java/1590.java
import java.util.*;

class Solution {
    public int minSubarray(int[] nums, int p) {
        int n = nums.length;
        int sum = 0;
        for (int i : nums) {
            sum = (sum + i) % p;
        }
        int target = sum % p;
        if (target == 0) {
            return 0;
        }

        Map<Integer, Integer> modMap = new HashMap<>();
        modMap.put(0, -1);

        int currentSum = 0;
        int result = n;
        for (int i = 0; i < n; i++) {
            currentSum = (currentSum + nums[i]) % p;
            int needed = (currentSum - target + p) % p;
            if (modMap.containsKey(needed)) {
                result = Math.min(result, i - modMap.get(needed));
            }
            modMap.put(currentSum, i);
        }
        return result == n ? -1 : result;
    }
}

File: Python/1590.py
from typing import List


class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n = len(nums)
        target = sum(nums) % p
        if target == 0:
            return 0

        mod_dict = {0: -1}
        curr_sum = 0
        min_len = n

        for i in range(n):
            curr_sum = (curr_sum + nums[i]) % p

            needed = (curr_sum - target + p) % p

            if needed in mod_dict:
                min_len = min(min_len, i - mod_dict[needed])

            mod_dict[curr_sum] = i

        return -1 if min_len == n else min_len
